File: 026-复杂链表的复制/026-linknext.cpp
/*
struct RandomListNode {
    int label;
    struct RandomListNode *next, *random;
    RandomListNode(int x) :
            label(x), next(NULL), random(NULL) {
    }
};
*/
class Solution {
public:
    RandomListNode* Clone(RandomListNode* pHead)
    {
        if(pHead==nullptr){
            return nullptr;
        }
        RandomListNode* cur=pHead;
        while(cur!=nullptr){
            RandomListNode* node=new RandomListNode(cur->label);
            node->next=cur->next;
            cur->next=node;
            cur=node->next;
        }
        cur=pHead;
        while(cur!=nullptr){
            if(cur->random!=nullptr){
                cur->next->random=cur->random->next;
            }
            cur=cur->next->next;
        }
        RandomListNode* c_head=pHead->next;
        RandomListNode* c_cur=c_head;
        cur=pHead;
        while(cur!=nullptr){
            cur->next=c_cur->next;
            cur=cur->next;
            if(c_cur->next==nullptr){
                break;
            }
            c_cur->next=cur->next;
            c_cur=c_cur->next;
        }
        return c_head;
    }
};

File: 026-复杂链表的复制/026-linknext.py
# -*- coding:utf-8 -*-
# class RandomListNode:
#     def __init__(self, x):
#         self.label = x
#         self.next = None
#         self.random = None
class Solution:
    # 返回 RandomListNode
    def Clone(self, pHead):
        if pHead is None:
            return None
        cur=pHead
        while cur:
            node=RandomListNode(cur.label)
            cur.next,node.next=node,cur.next
            cur=node.next
        cur=pHead
        while cur:
            if cur.random:
                cur.next.random=cur.random.next
            cur=cur.next.next
        c_head=pHead.next
        c_cur=c_head
        cur=pHead
        while cur:
            cur.next=c_cur.next
            cur=c_cur.next
            if c_cur.next is None:
                break
            c_cur.next=cur.next
            c_cur=cur.next
        return c_head
